frontend/udp/frontend.go

```diff
@@ -128,10 +128,15 @@ func (t *Frontend) ListenAndServe() error {
 			defer t.wg.Done()
 			defer pool.Put(buffer)
 
+			if ip := addr.IP.To4(); ip != nil {
+				addr.IP = ip
+			}
+
 			// Handle the request.
 			start := time.Now()
 			action, err := t.handleRequest(
-				Request{buffer[:n], addr.IP},
+				// Make sure the IP is copied, not referenced.
+				Request{buffer[:n], append([]byte{}, addr.IP...)},
```

Comment:
Apparently `[]byte(addr.IP)` does not make a copy of the underlying array, but we need it to be a copy.

If it's not a copy, the changes introduced in parser.go will overwrite the bytes of `addr.IP` which will cause the response packet to be sent to whatever address the client spoofed.

See this link on the playground: https://play.golang.org/p/0NdtJU_jqX